File: 3939-process-string-with-special-operations-ii/process-string-with-special-operations-ii.py
class Solution:
    def processStr(self, s: str, k: int) -> str:
        ans = 0

        for ch in s:
            if ch == "*":
                if ans:
                    ans -= 1
            elif ch == "#":
                ans *= 2
            elif ch == "%":
                continue
            else:
                ans += 1

        if ans <= k:
            return "."

        s = s[::-1]

        for ch in s:
            if ch == "*":
                ans += 1
            elif ch == "#":
                half = ans // 2
                if k >= half:
                    k -= half
                ans = half
            elif ch == "%":
                k = ans - k - 1
            else:
                if k == ans - 1:
                    return ch
                ans -= 1

        return "."

File: 3939-process-string-with-special-operations-ii/process-string-with-special-operations-ii.go
func processStr(s string, k int64) byte {
	var ans int64

	for i := 0; i < len(s); i++ {
		switch s[i] {
		case '*':
			if ans > 0 {
				ans--
			}
		case '#':
			ans *= 2
		case '%':
			continue
		default:
			ans++
		}
	}

	if k+1 > ans {
		return '.'
	}

	for i := len(s) - 1; i >= 0; i-- {
		switch s[i] {
		case '*':
			ans++
		case '#':
			half := ans / 2
			if k >= half {
				k -= half
			}
			ans = half
		case '%':
			k = ans - k - 1
		default:
			if k+1 == ans {
				return s[i]
			}
			ans--
		}
	}

	return '.'
}